xip/xipether.c

```diff
@@ -17,13 +17,116 @@
 #include "libnetlink.h"
 #include "xiart.h"
 
+#define IFINDEX_STR_SIZE 8
+#define XIA_LLADDR_LEN 12
+#define APPEND_ZEROES 20
+
 static int usage(void)
 {
 	fprintf(stderr,
-"Usage:	\n");
+"Usage:	xip ether { addinterface | delinterface } dev DEV\n"
+"       xip ether { addneigh | delneigh } lladdr LLADDR dev DEV\n"
+"       xip ether show { interfaces | neighs }\n"
+"where	LLADDR := HEXDIGIT{1,2} (':' HEXDIGIT{1,2})*\n"
+"	DEV := STRING NUMBER\n");
 	return -1;
 }
 
+static void form_ether_xid(unsigned oif, unsigned char *lladdr, unsigned tlen, char *id)
+{
+	char bytes[4];
+	unsigned append = APPEND_ZEROES;
+	bytes[0] = (oif >> 24) & 0xFF;
+	bytes[1] = (oif >> 16) & 0xFF;
+	bytes[2] = (oif >> 8) & 0xFF;
+	bytes[3] = oif & 0xFF;
+	snprintf(id, (tlen+1), "%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%0*hu",(unsigned char)bytes[0],
+				(unsigned char)bytes[1],(unsigned char)bytes[2],
+				(unsigned char)bytes[3],lladdr[0],lladdr[1],
+				lladdr[2], lladdr[3],lladdr[4], lladdr[5], append, 0);
+}
+
+static int modify_local(const struct xia_xid *dst, int to_add)
+{
+	struct {
+		struct nlmsghdr 	n;
+		struct rtmsg 		r;
+		char   			buf[1024];
+	} req;
+
+	memset(&req, 0, sizeof(req));
+
+	req.n.nlmsg_len = NLMSG_LENGTH(sizeof(struct rtmsg));
+
+	if (to_add) {
+		/* XXX Does one really needs all these flags? */
+		req.n.nlmsg_flags = NLM_F_REQUEST|NLM_F_CREATE|NLM_F_EXCL;
+		req.n.nlmsg_type = RTM_NEWROUTE;
+	} else {
+		req.n.nlmsg_flags = NLM_F_REQUEST;
+		req.n.nlmsg_type = RTM_DELROUTE;
+	}
+
+	req.r.rtm_family = AF_XIA;
+	req.r.rtm_table = XRTABLE_LOCAL_INDEX;
+	req.r.rtm_protocol = RTPROT_BOOT;
+	req.r.rtm_type = RTN_LOCAL;
+	req.r.rtm_scope = RT_SCOPE_HOST;
+
+	req.r.rtm_dst_len = sizeof(*dst);
+	addattr_l(&req.n, sizeof(req), RTA_DST, dst, sizeof(*dst));
+
+	if (rtnl_talk(&rth, &req.n, 0, 0, NULL, NULL, NULL) < 0)
+		exit(2);
+	return 0;
+}
+
+static int do_local(int argc, char **argv, int to_add)
+{
+	struct xia_xid dst;
+	const char *dev;
+	char strid[2*XIA_XID_MAX];
+	unsigned char lladdr[MAX_ADDR_LEN];
+	unsigned oif,addrlen;
+
+	if (argc != 2) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+
+	if (strcmp(argv[0], "dev")) {
+		fprintf(stderr, "Wrong parameters\n");
+		return usage();
+	}
+
+	dev = argv[1];
+	oif = ll_name_to_index(dev);
+	if (!oif) {
+		fprintf(stderr, "Cannot find device '%s'\n", dev);
+		return -1;
+	}
+	addrlen = ll_index_to_addr(oif, lladdr, sizeof(lladdr));
+	if (!addrlen) {
+		/* should add a check to see if addrlen also equal to 12?*/
```

Comment:
I think we don't need that check, since the function ``ll_index_to_addr()`` deals with different situations.